Frontend locations - Map with adjustable radius

lucaslcode opened the pull request:
First commit adds two components to storybook. Please review!

Mobile emulation in Edge wasn't allowing me to change the radius. I have to try on my phone somehow.

lucaslcode committed:
Add Map and EditLocationMap components to storybook

lucaslcode commented on app/frontend/src/components/EditLocationMap.tsx:
```diff
@@ -0,0 +1,132 @@
+
+  const center = useMemo(() => new LatLng(user.lat, user.lng), [user]);
+  //What type is this useRef supposed to be!? Marker doesn't want to work
+  const markerRef = useRef<any>(null);
```
Couldn't get typing to work here

amdwit replied:
Me neither. I think the correct typing is `typeof Marker` but then in next lines `getLatLng` is not accepted by Typescript.

lucaslcode committed:
Don't update location on map move

lucaslcode commented:
Also, I get lines between the tiles, not sure why. 
![image](https://user-images.githubusercontent.com/32044095/101104261-a8643880-35c2-11eb-8c85-4f294ed609db.png)

amdwit commented on app/frontend/src/components/Map.stories.tsx:
```diff
@@ -0,0 +1,39 @@
+    <p>An explicitly sized Map.</p>
+    <div style={{ width: "400px", border: "1px solid black" }}>
+      <Map {...args} grow />
+    </div>
```
I do not think it is necessary to include all these possibilities. If it covers the use cases as in the design and the website it is ok. I think you copied this from the Avatar where these different uses grew naturally based on how we want to use it. But I am not saying I would like to have some deleted here.

amdwit commented:
> Also, I get lines between the tiles, not sure why.
> ![image](https://user-images.githubusercontent.com/32044095/101104261-a8643880-35c2-11eb-8c85-4f294ed609db.png)

I do not see those on my browsers even not on Edge for Mac. Maybe it is browser dependent. Maybe it has to do with pixel calculation rounding.

lucaslcode committed:
Slim storybook maps

lucaslcode committed:
Add MapSearch leaflet component

lucaslcode committed:
Take react searchbox out of leaflet and overlay it

lucaslcode committed:
Start switch to react-map-gl

lucaslcode committed:
Make editable radius marker

I am going to add comments and clean it up later. This took bloody forever and required workaround after workaround.

At first I wanted to drag the circle anywhere to resize it. However the library doesn't support this, you can only drag to move, or it tries to move and resize at the same time. So I had to go with a second small "handle" which can be moved to resize.

However it's still not perfect, moving the circle resets the marker "angle" back to 0 degrees.

lucaslcode committed:
Finish Map for editing an approximate Location

now with comments and double-click to set

lucaslcode committed:
Continue frontend Map

This is nearly working but it's a buggy mess. The React-Map-GL bindings are too limiting, next commit I am going to ditch them and just use declarative syntax and a geojson marker instead. Will be much smoother.

lucaslcode committed:
start switch to imperative mapbox

adjustible radius marker working except handle and circle both receive click events on an overlapping click

lucaslcode committed:
Complete draggable radius

Mapbox GL JS has a problem where clicks on overlapping layers fires events for all the layers. It is possible to check if the event has fired before, but not possible to choose the order. In this case, the circle was always firing before the marker/resize handle, when the opposite would be is important for UX.

As a workaround, I offset the resize handle so there is no overlap, but this makes the code rather verbose. I don't see another way around this but I would love to hear any possibilities.

lucaslcode committed:
Cleanup - Complete adjustable radius

The radius has a handle to resize which is offset from the circle. This makes the code a bit messy, but was a necessary workaround because overlapping elements ALL fire their mouse events and it messed everything up. It is possible to check if the event has been fired, but the circle event was always firing first, when it should have been the handle event for good UX.

lucaslcode committed:
pin mapbox-gl at <2.0.0

lucaslcode commented:
I will split this into multiple PRs as this is already a monster. So far just adds the map with an adjustable radius to storybook.
I have pinned mapbox-gl to <2.0.0 as v2 requires an api key (even though we are indeed using one, for now).

PS. look at the commit history if you want to feel my pain

amdwit commented:
I do not see a marker in Storybook/Components/Map/Primary top map

lucaslcode commented:
@amdwit sorry markers got lost in refactoring and I haven't added any back in yet

lucaslcode committed:
Merge branch 'develop' into frontend/feature/user-locations
